channeldb/invoices.go

```diff
@@ -538,7 +553,9 @@ func (d *DB) InvoicesAddedSince(sinceAddIndex uint64) ([]Invoice, error) {
 // full invoice is returned. Before setting the incoming HTLC, the values
 // SHOULD be checked to ensure the payer meets the agreed upon contractual
 // terms of the payment.
-func (d *DB) LookupInvoice(paymentHash [32]byte) (Invoice, error) {
+func (d *DB) LookupInvoice(paymentHash [32]byte,
+	payAddr *[32]byte) (Invoice, error) {
```

Comment:
Is there a use case for specifying both?

Reply:
Or if both are specified, is a sanity check useful to see if hash is always present and pointing to the same invoice if addr is present too?

Reply:
the reason to use both is so that we don't have to add a migration for existing invoices. with AMP the hash won't align with any particular invoice, but this wil be worked around by providing a nil hash in the `InvoiceRef`.

if we do a full migration we'd end up supporting either `(hash, nil)` or `(nil, pay_addr)`. if we don't, then we need to support `(hash, nil)`, `(nil, pay_addr)`, or (hash, pay_addr)`, where the latter has a sanity check that they point to the same invoice.

atm, there isn't really a business use case for doing the full migration, it would only serve to clean up this bit of code. this may change in the in the future tho (if we begin exposing invoice lookups by payment addr and we want to expose these slightly older invoices), but it is still possible to do this migration retroactively when that time comes. it is not terribly difficult, but it is time consuming in the critical path

Reply:
`InvoiceRef` added, supporting `(hash, nil)` and `(hash, pay_addr)`. the final variant will be added once the related AMP changes go in